crates/viewer/re_renderer/src/video/player.rs

```diff
@@ -439,71 +481,68 @@ impl VideoPlayer {
+    if decoded_frame_pts == requested_pts {
+        // Decoder caught up with request!
+        *dont_update_texture_until_caught_up = false;
+        true
+    } else {
+        // Decoder did not produce the desired frame, but something _before_ the requested frame.
+        if *dont_update_texture_until_caught_up {
+            // We're waiting until we're fully caught up before showing any new frames.
+            false
+        } else {
+            // Figure out how many frames we're behind. If this is higher than a certain tolerance, don't update the texture.
+            //
+            // Since frames aren't in presentation time order and may have varying durations (i.e. the video has variable frame rate),
+            // we have to successively use `latest_sample_index_at_presentation_timestamp`:
+            // Start at the desired sample and walk backwards from there until we find the sample for the actually produced frame.
+            let mut num_frames_behind = 0;
+            let mut sample = requested_sample;
+            loop {
+                let Some(current_sample) = sample else {
+                    // Sample doesn't exist anymore. Nothing we can do!
+                    break false;
+                };
+                if current_sample.presentation_timestamp == decoded_frame_pts {
+                    // This is the frame we actually got and we stayed under the tolerance.
+                    // This may happen if the load on the decoder fluctuates or it is just about able to keep up with playback.
+                    break true;
+                }
 
-                // Might this be due to an error?
-                //
-                // We only care about decoding errors when we don't find the requested frame,
-                // since we want to keep playing the video fine even if parts of it are broken.
-                // That said, practically we reset the decoder and thus all frames upon error,
-                // so it doesn't make a lot of difference.
-                if let Some(timed_error) = last_error {
-                    if DECODING_GRACE_DELAY <= timed_error.time_of_first_error.elapsed() {
-                        // Report the error only if we have been in an error state for a certain amount of time.
-                        // Don't immediately report the error, since we might immediately recover from it.
-                        // Otherwise, this would cause aggressive flickering!
-                        return Err(timed_error.latest_error.clone());
-                    }
+                num_frames_behind += 1;
+                if num_frames_behind > TOLERATED_OUTPUT_DELAY_IN_NUM_FRAMES {
+                    // Too far behind. Add hysteresis to avoid flipping in and out of this state.
+                    *dont_update_texture_until_caught_up = true;
+                    break false;
```

Comment:
This function is effectively setting two bools (`dont_update_texture_until_caught_up` and the `return` value). 

Do we really need the return value? Let's rename the state from `dont_update_texture_until_caught_up` to `is_catching_up`.

* If we are caught up, `is_catching_up` is `false` and we should update the texture
* If we're far behind, `is_catching_up` is `true` and we should not update the texture
* If we are only a bit behind, then we should only update the texture if `is_catching_up == false.

So the decision on updating the texture is exactly the the state of `is_catching_up`?

The only corner cases I see is when a sample is missing or similar.
I think those cases can still fit under a single `bool` state (with an appropriate name) or a tri-state enum.